core/src/kura.rs

```diff
@@ -4,258 +4,475 @@
+    ) -> Result<()>;
+    /// Write a single block index at the specified `block_height`.
+    /// If `block_height` is beyond the end of the index file attempt to
+    /// extend the index file.
+    /// # Errors
+    /// IO Error.
+    fn write_block_index(&mut self, block_height: u64, start: u64, length: u64) -> Result<()>;
 
-    async fn handle(&mut self, StoreBlock(block): StoreBlock) {
-        #[cfg(feature = "telemetry")]
-        if block.header().height == 1 {
-            iroha_logger::telemetry!(msg = iroha_telemetry::msg::SYSTEM_CONNECTED, genesis_hash = %block.hash());
-        }
-        if let Err(error) = self.store(block).await {
-            error!(%error, "Failed to write block")
+    /// Read the amount of indices in the index file. (Byte size of
+    /// the index file divided by 16 B.)
+    /// # Errors
+    /// IO Error. Because this function
+    /// does so little, if this fails you can be quite sure all other
+    /// read and write ops will fail. The most common reason this function
+    /// errors is because you forgot to call `create_files_if_they_do_not_exist`.
+    fn read_index_count(&self) -> Result<u64>;
+    /// Change the value returned by `read_index_count` aka the size of the
+    /// index file.
+    /// # Errors
+    /// IO Error.
+    fn write_index_count(&mut self, new_count: u64) -> Result<()>;
+
+    /// Read from the data file in order to fill `dest_buffer` starting
+    /// from `start_location_in_data_file`.
+    /// # Errors
+    /// IO Error.
+    fn read_block_data(
+        &self,
+        start_location_in_data_file: u64,
+        dest_buffer: &mut [u8],
+    ) -> Result<()>;
+    /// Write `block_data` into the data file starting at
+    /// `start_location_in_data_file`. Extend the file if
+    /// necessary.
+    /// # Errors
+    /// IO Error.
+    fn write_block_data(
+        &mut self,
+        start_location_in_data_file: u64,
+        block_data: &[u8],
+    ) -> Result<()>;
+
+    /// Create the index and data files if they do not
+    /// already exist.
+    /// # Errors
+    /// Will fail if the any of the files don't exist
+    /// and couldn't be created.
+    fn create_files_if_they_do_not_exist(&mut self) -> Result<()>;
+
+    /// Above are the platform dependent functions.
+    /// Below are the platform independent functions that use the above functions.
+
+    /// Short hand for calling `read_block_indices` with a buffer of one.
+    /// # Errors
+    /// Passes on any errors from `read_block_indices`.
+    fn read_block_index(&self, block_height: u64) -> Result<(u64, u64)> {
+        let mut index = (0, 0);
+        self.read_block_indices(block_height, std::slice::from_mut(&mut index))?;
+        Ok(index)
+    }
+
+    /// Append `block_data` to this block store. First writing
+    /// the data to the data file and then creating a new index
+    /// for it.
+    /// # Errors
+    /// Fails if any of the required platform specific function operations
+    /// fail.
```

Comment:
```suggestion
    /// Append `block_data` to this block store: write
    /// the data to the data file and then create a new index
    /// for it.
    ///
    /// # Errors
    /// Fails if any of the required platform-specific function operations
    /// fail.
```

`function operations` -- maybe just `functions` or just `operations`?